File: impl/ts/tests/interval.test.ts
// Cross-check: the TS interval parser/renderer reproduces the byte-exact vectors in
// spec/encoding/intervals.toml (CLAUDE.md §8) — identical to the Rust/Go cores. All cascade/span
// math is bigint (number loses i64 precision).

import assert from "node:assert/strict";
import { test } from "node:test";
import { EngineError } from "../src/errors.ts";
import {
  type Interval,
  intervalCmp,
  intervalEncodeKey,
  intervalSpan,
  parseInterval,
  renderInterval,
} from "../src/interval.ts";
import { readIntervalCases, specPath } from "./tomlmini.ts";

const hex = (b: Uint8Array) => Array.from(b, (x) => x.toString(16).padStart(2, "0")).join("");
const cmpBytes = (a: Uint8Array, b: Uint8Array) => {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) if (a[i] !== b[i]) return a[i]! - b[i]!;
  return a.length - b.length;
};

test("interval vectors (parse/render byte-identical to Rust/Go)", () => {
  const cases = readIntervalCases(specPath("encoding/intervals.toml"));
  assert.ok(cases.length > 0, "no interval vectors parsed");
  for (const c of cases) {
    if (c.section === "parse") {
      const got = parseInterval(c.fields.input!);
      assert.equal(
        got.months,
        Number(c.fields.months),
        `parse ${JSON.stringify(c.fields.input)} months`,
      );
      assert.equal(got.days, Number(c.fields.days), `parse ${JSON.stringify(c.fields.input)} days`);
      assert.equal(
        got.micros,
        BigInt(c.fields.micros!),
        `parse ${JSON.stringify(c.fields.input)} micros`,
      );
    } else if (c.section === "parse_error") {
      let code = "";
      try {
        parseInterval(c.fields.input!);
      } catch (e) {
        if (e instanceof EngineError) code = e.code();
        else throw e;
      }
      assert.equal(code, c.fields.error, `parse ${JSON.stringify(c.fields.input)} error`);
    } else {
      const iv: Interval = {
        months: Number(c.fields.months),
        days: Number(c.fields.days),
        micros: BigInt(c.fields.micros!),
      };
      assert.equal(renderInterval(iv), c.fields.text, `render ${JSON.stringify(c.fields)}`);
    }
  }
});

test("interval span is canonical (1 mon == 30 days == 720:00:00)", () => {
  const oneMonth = parseInterval("1 mon");
  const thirtyDays = parseInterval("30 days");
  const hours = parseInterval("720:00:00");
  assert.equal(intervalCmp(oneMonth, thirtyDays), 0);
  assert.equal(intervalCmp(oneMonth, hours), 0);
  // span-equal but render distinctly
  assert.notEqual(renderInterval(oneMonth), renderInterval(thirtyDays));
  assert.ok(intervalCmp(parseInterval("1 day"), parseInterval("2 days")) < 0);
  assert.ok(intervalCmp(parseInterval("-1 day"), parseInterval("1 day")) < 0);
});

// The order-preserving KEY body (interval-span-i128, encoding.md §2.10): the 16-byte i128 span
// (bias 2^127 + big-endian). Sorting by intervalEncodeKey must equal span order; span-equal
// intervals share a key (the "equal but not identical" UNIQUE wrinkle, decimal's 1.5/1.50);
// byte-exact against the canonical vectors (spec/encoding/interval.toml).
test("interval encodeKey is order-preserving and byte-exact", () => {
  const iv = (months: number, days: number, micros: bigint): Interval => ({
    months,
    days,
    micros,
  });
  // Ascending by span — sorting by key must reproduce this order (sign boundary, zero, ±µs).
  const ordered = [
    iv(-1200, 0, 0n),
    iv(-1, 0, 0n),
    iv(0, -1, 0n),
    iv(0, 0, -1_000_000n),
    iv(0, 0, -1n),
    iv(0, 0, 0n),
    iv(0, 0, 1n),
    iv(0, 0, 1_000_000n),
    iv(0, 1, 0n),
    iv(1, 0, 0n),
    iv(1200, 0, 0n),
  ];
  const byKey = [...ordered].sort((a, b) => cmpBytes(intervalEncodeKey(a), intervalEncodeKey(b)));
  for (let i = 0; i < ordered.length; i++) {
    assert.equal(
      intervalSpan(byKey[i]!),
      intervalSpan(ordered[i]!),
      `encode_key order must equal span order at ${i}`,
    );
  }
  // Span-equal intervals share a key (1 mon == 30 days == 720:00:00) — the UNIQUE wrinkle.
  assert.equal(hex(intervalEncodeKey(iv(1, 0, 0n))), hex(intervalEncodeKey(iv(0, 30, 0n))));
  assert.equal(
    hex(intervalEncodeKey(iv(1, 0, 0n))),
    hex(intervalEncodeKey(iv(0, 0, 30n * 86_400_000_000n))),
  );
  // Byte-exact canonical vectors (the §2.10 worked-bytes table).
  assert.equal(hex(intervalEncodeKey(iv(0, 0, 0n))), "80000000000000000000000000000000");
  assert.equal(hex(intervalEncodeKey(iv(0, 0, 1n))), "80000000000000000000000000000001");
  assert.equal(hex(intervalEncodeKey(iv(0, 0, -1n))), "7fffffffffffffffffffffffffffffff");
  assert.equal(hex(intervalEncodeKey(iv(0, 1, 0n))), "8000000000000000000000141dd76000");
  assert.equal(hex(intervalEncodeKey(iv(1, 0, 0n))), "80000000000000000000025b7f3d4000");
  assert.equal(hex(intervalEncodeKey(iv(0, -1, 0n))), "7fffffffffffffffffffffebe228a000");
});

File: impl/go/interval_test.go
package jed

// Cross-check: the Go interval parser/renderer reproduces the byte-exact vectors in
// spec/encoding/intervals.toml (CLAUDE.md §8) — identical to the Rust/TS cores. Reuses the tiny
// inline-table scanner helpers from timestamp_test.go (same package). Test-only.

import (
	"bytes"
	"encoding/hex"
	"os"
	"sort"
	"strconv"
	"strings"
	"testing"
)

type ivCase struct {
	section string // "parse" | "parse_error" | "render"
	fields  map[string]string
}

func readIntervalCases(t *testing.T, path string) []ivCase {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var out []ivCase
	section := ""
	for _, raw := range strings.Split(string(data), "\n") {
		line := strings.TrimSpace(raw)
		switch {
		case line == "[[parse]]":
			section = "parse"
		case line == "[[parse_error]]":
			section = "parse_error"
		case line == "[[render]]":
			section = "render"
		case strings.HasPrefix(line, "{") && section != "":
			out = append(out, ivCase{section: section, fields: parseInlineFields(line)})
		}
	}
	return out
}

func TestIntervalVectors(t *testing.T) {
	t.Parallel()
	cases := readIntervalCases(t, specPath(t, "encoding/intervals.toml"))
	if len(cases) == 0 {
		t.Fatal("no interval vectors parsed")
	}
	for _, c := range cases {
		switch c.section {
		case "parse":
			in := unquote(c.fields["input"])
			months, _ := strconv.ParseInt(c.fields["months"], 10, 32)
			days, _ := strconv.ParseInt(c.fields["days"], 10, 32)
			micros, _ := strconv.ParseInt(c.fields["micros"], 10, 64)
			got, err := parseInterval(in)
			switch {
			case err != nil:
				t.Errorf("parse %q: unexpected error %v", in, err)
			case got.Months != int32(months) || got.Days != int32(days) || got.Micros != micros:
				t.Errorf("parse %q = {%d,%d,%d}, want {%d,%d,%d}", in,
					got.Months, got.Days, got.Micros, months, days, micros)
			}
		case "parse_error":
			in := unquote(c.fields["input"])
			want := unquote(c.fields["error"])
			_, err := parseInterval(in)
			ee, ok := err.(*EngineError)
			if !ok {
				t.Errorf("parse %q: expected error %s, got %v", in, want, err)
			} else if ee.Code() != want {
				t.Errorf("parse %q: error %s, want %s", in, ee.Code(), want)
			}
		case "render":
			months, _ := strconv.ParseInt(c.fields["months"], 10, 32)
			days, _ := strconv.ParseInt(c.fields["days"], 10, 32)
			micros, _ := strconv.ParseInt(c.fields["micros"], 10, 64)
			want := unquote(c.fields["text"])
			iv := Interval{Months: int32(months), Days: int32(days), Micros: micros}
			if got := renderInterval(iv); got != want {
				t.Errorf("render {%d,%d,%d} = %q, want %q", months, days, micros, got, want)
			}
		}
	}
}

func TestIntervalSpanCanonical(t *testing.T) {
	t.Parallel()
	oneMonth, _ := parseInterval("1 mon")
	thirtyDays, _ := parseInterval("30 days")
	hours, _ := parseInterval("720:00:00")
	if oneMonth.SpanCmp(thirtyDays) != 0 || oneMonth.SpanCmp(hours) != 0 {
		t.Errorf("'1 mon', '30 days', '720:00:00' should be span-equal")
	}
	if renderInterval(oneMonth) == renderInterval(thirtyDays) {
		t.Errorf("span-equal intervals should still render distinctly")
	}
	day, _ := parseInterval("1 day")
	twoDays, _ := parseInterval("2 days")
	negDay, _ := parseInterval("-1 day")
	if day.SpanCmp(twoDays) >= 0 || negDay.SpanCmp(day) >= 0 {
		t.Errorf("span ordering wrong")
	}
}

// TestIntervalEncodeKey verifies the order-preserving KEY body (interval-span-i128,
// encoding.md §2.10): the 16-byte i128 span (bias 2^127 + big-endian). Sorting by EncodeKey must
// equal span order; span-equal intervals share a key (the "equal but not identical" UNIQUE
// wrinkle, decimal's 1.5/1.50); byte-exact against the canonical vectors (interval.toml).
func TestIntervalEncodeKey(t *testing.T) {
	t.Parallel()
	iv := func(m, d int32, u int64) Interval { return Interval{Months: m, Days: d, Micros: u} }
	// Ascending by span — sorting by key must reproduce this order (sign boundary, zero, ±µs).
	ordered := []Interval{
		iv(-1200, 0, 0), iv(-1, 0, 0), iv(0, -1, 0), iv(0, 0, -1_000_000), iv(0, 0, -1),
		iv(0, 0, 0), iv(0, 0, 1), iv(0, 0, 1_000_000), iv(0, 1, 0), iv(1, 0, 0), iv(1200, 0, 0),
	}
	byKey := append([]Interval(nil), ordered...)
	sort.Slice(byKey, func(i, j int) bool {
		return bytes.Compare(byKey[i].EncodeKey(), byKey[j].EncodeKey()) < 0
	})
	for i := range ordered {
		if byKey[i].SpanCmp(ordered[i]) != 0 {
			t.Fatalf("encode_key order must equal span order at %d", i)
		}
	}
	// Span-equal intervals share a key (1 mon == 30 days == 720:00:00) — the UNIQUE wrinkle.
	if !bytes.Equal(iv(1, 0, 0).EncodeKey(), iv(0, 30, 0).EncodeKey()) {
		t.Errorf("1 mon / 30 days keys must coincide")
	}
	if !bytes.Equal(iv(1, 0, 0).EncodeKey(), iv(0, 0, 30*86_400_000_000).EncodeKey()) {
		t.Errorf("1 mon / 720:00:00 keys must coincide")
	}
	// Byte-exact canonical vectors (the §2.10 worked-bytes table).
	exact := []struct {
		iv   Interval
		want string
	}{
		{iv(0, 0, 0), "80000000000000000000000000000000"},
		{iv(0, 0, 1), "80000000000000000000000000000001"},
		{iv(0, 0, -1), "7fffffffffffffffffffffffffffffff"},
		{iv(0, 1, 0), "8000000000000000000000141dd76000"},
		{iv(1, 0, 0), "80000000000000000000025b7f3d4000"},
		{iv(0, -1, 0), "7fffffffffffffffffffffebe228a000"},
	}
	for _, e := range exact {
		if got := hex.EncodeToString(e.iv.EncodeKey()); got != e.want {
			t.Errorf("EncodeKey(%+v) = %s, want %s", e.iv, got, e.want)
		}
	}
}
